nixos/zfs: Ensure pool has datasets to decrypt

ElvishJerricco opened the pull request:
###### Description of changes

This should fix #203083 and fix #202164. The ZFS module was incorrectly emitting `zfs list -Ho name,keylocation,keystatus` with no dataset argument in cases where `requestEncryptionCredentials` is a list that doesn't have at least one dataset from each pool.

###### Things done

~~I haven't tested this yet, so I need to write a test for it.~~ Done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ElvishJerricco committed:
nixos/zfs: Ensure pool has datasets to decrypt

ElvishJerricco commented:
@ofborg test zfs

ElvishJerricco commented:
@ofborg test zfs

winterqt reviewed (commented):
Is there any particular reason you used `nixos: ...` in the commit message instead of `nixos/zfs`?

ElvishJerricco commented:
@winterqt Nope, but that's a better choice.

winterqt commented on nixos/tests/zfs.nix:
```diff
@@ -24,96 +24,144 @@ let
 
-        environment.systemPackages = [ pkgs.parted ];
+        specialisation.encryption.configuration = {
+          boot.zfs.requestEncryptionCredentials = ["automatic"];
```
Please use consistent formatting in newly-added code (you're using spaces in some arrays and not others, it seems).

winterqt commented on nixos/tests/zfs.nix:
```diff
@@ -24,96 +24,144 @@ let
             device = "forcepool";
             fsType = "zfs";
-            options = [ "noauto" ];
+            options = ["noauto"];
```
Maybe consider undoing this change, to keep the diff focused? Probably a nit.

ElvishJerricco replied:
I ended up just changing all the lists in the diff to have the spaces. That way I could undo this change and still have everything using consistent style.

winterqt commented on nixos/tests/zfs.nix:
```diff
@@ -24,96 +24,144 @@ let
       };
 
       testScript = ''
+        machine.wait_for_unit("multi-user.target")
```
Why is this suddenly needed?

ElvishJerricco replied:
@winterqt Previously, it ran `udevadm settle` to wait for the time to format the drive. I find it cleaner to just wait for the system to be booted.

winterqt reviewed (commented):
This looks good overall, just a few comments. Sorry for taking so long to get to this, and let me know if any of them are overstepping.

winterqt commented on nixos/tests/zfs.nix:
```diff
@@ -17,103 +17,151 @@ let
                 "zgenhostid deadcafe",
-                "zpool create forcepool /dev/vdb1 -O mountpoint=legacy",
+                "zpool create forcepool /dev/vdc1 -O mountpoint=legacy",
+                "bootctl set-default nixos-generation-1-specialisation-forcepool.conf",
```
Or this one, for that matter.

ElvishJerricco replied:
If you mean the change from `vdb` to `vdc`, that's because `virtualisation.useBootLoader = true;` adds another disk to the scenario, and the disks we actually want to use become `vdc` and `vdd` instead.

If you mean the `bootctl` command, that's part of why we need `useBootLoader` now. The test better reflects real world operation if some things are done as `specialisation` configurations instead of manually issuing commands to the same effect. This requires boot loader configs that we can set as the default.

winterqt reviewed (changes requested):
Firstly, with regards to the commit messages:

```diff
- nixos/zfs: Ensure pool has datasets to decrypt
+ nixos/zfs: ensure pool has datasets to decrypt

(Maybe adding a description similar to the PR description would be a good idea? *shrugs*)

- nixos/tests/zfs: Test requestEncryptionCredentials as a list.
+ nixos/tests/zfs: test requestEncryptionCredentials as a list
```

Secondly, I'm admittedly a bit confused with some of the changes in the test... would it make sense to split them into the requestEncryptionCredentials changes, and then the Samba ones? Or however else it would make sense?

Let me know if these suggestions are misguided. 😅

ElvishJerricco commented on nixos/tests/zfs.nix:
```diff
@@ -17,103 +17,151 @@ let
+                "bootctl set-default nixos-generation-1-specialisation-encryption.conf",
+                "sync",
+                "zpool export automatic",
+                "zpool export manual",
+            )
+            machine.crash()
+            machine.start()
+            machine.wait_for_console_text("Enter key for")
+            machine.send_console("password\n")
+            machine.wait_for_unit("multi-user.target")
+            machine.succeed(
+                "zfs get keystatus manual/encrypted | grep unavailable",
+                "echo otherpass | zfs load-key manual/encrypted",
+                "systemctl start manual-encrypted.mount",
+                "umount /automatic /manual/encrypted /manual",
+                "zpool destroy automatic",
+                "zpool destroy manual",
+            )
 
         with subtest("boot.zfs.forceImportAll works"):
             machine.succeed(
                 "rm /etc/hostid",
                 "zgenhostid deadcafe",
-                "zpool create forcepool /dev/vdb1 -O mountpoint=legacy",
+                "zpool create forcepool /dev/vdc1 -O mountpoint=legacy",
+                "bootctl set-default nixos-generation-1-specialisation-forcepool.conf",
+                "rm /etc/hostid",
+                "sync",
             )
-            machine.shutdown()
-            machine.start()
-            machine.succeed("udevadm settle")
+            machine.crash()
+            machine.wait_for_unit("multi-user.target")
             machine.fail("zpool import forcepool")
             machine.succeed(
-                "systemctl start zfs-import-forcepool.service",
-                "mount -t zfs forcepool /tmp/mnt",
+                "systemctl start forcepool.mount",
+                "mount | grep forcepool",
```
This was mostly to ensure the test properly reflects the way `systemd` will cause the mount and the import service to be started and ordered. `systemctl start foo.mount` will properly invoke other units while `mount /foo` will not.

ElvishJerricco commented:
@winterqt Admittedly, most of the changes to the test are more about cleaning it up and making it more representative of real world use. I think those changes were necessary in order to add testing for what I actually wanted to test, but I suppose it could possibly have come in the form of a separate commit.

winterqt commented:
> Let me know if you would like the more generic changes split out into their own commit.

I'd appreciate it just to make the changes more clear, yeah -- unless you think I'm being too picky here.

ElvishJerricco committed:
nixos/tests/zfs: Represent real world usage better

It's better to utilize the boot process and systemd mechanisms to test
these zfs features, rather than manually simulating the same behavior
with testScript.

ElvishJerricco committed:
nixos/tests/zfs: Test requestEncryptionCredentials as a list.

ElvishJerricco commented:
@winterqt done.

ElvishJerricco commented:
@ofborg test zfs

winterqt commented:
Split looks good, thanks.

I guess the only thing I'd want as an outsider is an explanation of what "real world usage" actually means. I don't know, maybe I'm just being too pedantic, but I think it may be good to explain what exactly these changes mean in terms of representing real world usage (as you've explained in e.g. https://github.com/NixOS/nixpkgs/pull/203171#discussion_r1045558437).

I'm sorry I'm being so stubborn (/shallow?) on this, and also for taking so long to respond to changes. Feel free to just ignore me by this point tbh 🙃

ElvishJerricco commented:
@winterqt I've updated the commit message to explain a bit better. Will merge once ofborg is happy.
